meta: turn this repo into a monorepo

ephys opened the pull request:
## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? <!-- Put PR link here -->
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Writing plugins & packages that depend on core is a headache without a proper monorepo setup. This PR sets that up.

This should make it trivial to create new packages such as [`@sequelize/postgres`](https://github.com/sequelize/sequelize/issues/13722), and [`@sequelize/validator.js`](https://github.com/sequelize/sequelize/issues/15497)

I'm sure I have missed something. Let's find out 🤞

I haven't tested the publish setup. If it fails, I'll open a follow-up PR to debug it

WikiRik commented:
Let's publish a new alpha on the latest channel after we merge this to check if it works.

Another thing we can't really check right now and might need to fix in a follow up PR is the behaviour of renovate

WikiRik commented on packages/validator-js/package.json:
```diff
@@ -0,0 +1,25 @@
+    "url": "https://github.com/sequelize/sequelize/issues"
+  },
+  "main": "index.js",
+  "author": "",
```
```suggestion
```

WikiRik replied:
There are some things we should add as well but I don't think they're for this PR.

ephys replied:
Yep, follow-up PR :)

WikiRik commented on packages/validator-js/package.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "@sequelize/validator.js",
```
Will this be released as well? Or does it start with only `@sequelize/core`?

ephys replied:
The goal is to publish a dummy release, to see if it works

WikiRik commented on packages/core/package.json:
```diff
@@ -0,0 +1,224 @@
+    "test-unit-ibmi": "cross-env DIALECT=ibmi yarn test-unit",
+    "test-unit-snowflake": "cross-env DIALECT=snowflake yarn test-unit",
+    "test-unit-all": "yarn test-unit-mariadb && yarn test-unit-mysql && yarn test-unit-postgres && yarn test-unit-mssql && yarn test-unit-sqlite && yarn test-unit-snowflake && yarn test-unit-db2 && yarn test-unit-ibmi",
+    "----------------------------------------- integration tests ---------------------------------------------": "",
```
```suggestion
    "---------------------------------------- integration tests --------------------------------------------": "",
```
Or however long we need it to be to the same size as the others

ephys replied:
Done

WikiRik commented on .github/workflows/ci.yml:
```diff
@@ -32,9 +32,11 @@ jobs:
+        run: yarn workspace @sequelize/core build
       - name: Compress artifact
-        run: tar -cf install-build-node-${{ matrix.node-version }}.tar ./lib ./node_modules ./types
+        run: tar -cf install-build-node-${{ matrix.node-version }}.tar ./packages/core/lib ./packages/core/node_modules ./packages/core/types ./node_modules
```
Which dependencies are in `./node_modules`?

WikiRik replied:
Considering that for the `typedoc` check we still install the root dependencies

ephys replied:
Yarn workspace dedupes dependencies needed by the different packages of the workspace by putting them in the root node_modules

Got to remove the redundant install from the docs script

ephys replied:
Done

WikiRik reviewed (commented):
Haven't really worked with `yarn workspaces` before and might need to get used to it a bit, especially with running the scripts using `yarn workspace @sequelize/core test-unit-all` for example. In a future PR we might want to look into making some scripts available on the root package as well (`test-unit-all` and `test-integration-{dialect}` comes to my mind first)

WikiRik reviewed (approved):
I think we should be able to test this out

ephys commented:
Had to merge the changes made by renovate. We'll need to merge this before renovate updates something again

Also changed how the CI installs node_modules: Running `yarn install` at the root of the repo installs the dependencies of all packages automatically. And removed the lockfile from `packages/core`, because yarn doesn't actually use it. It uses the one at the root of the repo

ephys commented:
🙌
